examples/qwen2.5_nemo/training/model/train.py

```diff
@@ -0,0 +1,187 @@
+import argparse
+import os
+import nemo_run as run
+import lightning.pytorch as pl
+from nemo.collections import llm
+
+import torch
+print(f"CUDA available: {torch.cuda.is_available()}")
+print(f"Number of GPUs: {torch.cuda.device_count()}")
+
+### Dataset
+from data import BespokeDataModule
+
+def bespoke(seq_length: int = 8000, micro_batch_size: int = 1, global_batch_size: int = 4, num_workers: int = 1) -> run.Config[pl.LightningDataModule]:
+    return run.Config(BespokeDataModule, seq_length=seq_length, micro_batch_size=micro_batch_size, global_batch_size=global_batch_size, num_workers=num_workers)
+
+import nemo_run as run
+from nemo import lightning as nl
+from nemo.collections import llm
+from megatron.core.optimizer import OptimizerConfig
+import torch
+import lightning.pytorch as pl
+from pathlib import Path
+from nemo.collections.llm.recipes.precision.mixed_precision import bf16_mixed
+
+# Configure the trainer
+def trainer(devices: int = 4, max_steps: int = 300, tensor_model_parallel_size: int = 4, log_every_n_steps: int = 1) -> run.Config[nl.Trainer]:
+    strategy = run.Config(
+        nl.MegatronStrategy,
+        tensor_model_parallel_size=tensor_model_parallel_size,
+    )
+    trainer = run.Config(
+        nl.Trainer,
+        devices=devices,
+        max_steps=max_steps,
+        accelerator="gpu",
+        strategy=strategy,
+        plugins=bf16_mixed(),
+        log_every_n_steps=log_every_n_steps,
+        limit_val_batches=0,
+        val_check_interval=0,
+        num_sanity_val_steps=0,
+    )
+    return trainer
+
+# Configure the logger
+def logger(name: str = "qwen_sft", log_dir: str = "./results", checkpoint_every_n_steps: int = 100, save_top_k: int = 1) -> run.Config[nl.NeMoLogger]:
+    ckpt = run.Config(
+        nl.ModelCheckpoint,
+        save_last=True,
+        every_n_train_steps=checkpoint_every_n_steps,
+        monitor="reduced_train_loss",
+        save_top_k=save_top_k,
+        save_on_train_epoch_end=True,
+        save_optim_on_train_end=True,
+    )
+
+    return run.Config(
+        nl.NeMoLogger,
+        name=name,
+        log_dir=log_dir,
+        use_datetime_version=False,
+        ckpt=ckpt,
+        wandb=None
+    )
+
+# Configure the optimizer
+def adam_with_cosine_annealing(lr: float = 2e-5, adam_beta2: float = 0.98, clip_grad: float = 1.0) -> run.Config[nl.OptimizerModule]:
+    opt_cfg = run.Config(
+        OptimizerConfig,
+        optimizer="adam",
+        lr=lr,
+        adam_beta2=adam_beta2,
+        use_distributed_optimizer=True,
+        clip_grad=clip_grad,
+        bf16=True,
+    )
+    return run.Config(
+        nl.MegatronOptimizerModule,
+        config=opt_cfg
+    )
+
+# Configure the model
+def qwen() -> run.Config[pl.LightningModule]:
+    return run.Config(llm.Qwen2Model, config=run.Config(llm.Qwen2Config7B))
+
+# Configure the resume
+def resume(model_id: str = "Qwen/Qwen2.5-7B-Instruct") -> run.Config[nl.AutoResume]:
```

Comment:
have we tested that this works with cache? If so, can you add a comment about expected behavior?